File: templates/body.go
package templates

import (
	"bytes"
	"fmt"
	"html"
	"regexp"
	"strconv"
	"strings"

	"github.com/bakape/meguca/common"
	"github.com/bakape/meguca/config"
	"github.com/valyala/quicktemplate"
)

// Embeddable URL types
const (
	youTube = iota
	soundCloud
	vimeo
)

var (
	commandRegexp   = regexp.MustCompile(`^#(flip|\d*d\d+|8ball|pyu|pcount)$`)
	diceRegexp      = regexp.MustCompile(`^(\d*)d(\d+)$`)
	linkRegexp      = regexp.MustCompile(`^>>(>*)(\d+)$`)
	referenceRegexp = regexp.MustCompile(`^>>>(>*)\/(\w+)\/$`)
	urlRegexp       = regexp.MustCompile(
		`^(?:magnet:\?|https?:\/\/)[-a-zA-Z0-9@:%_\+\.~#\?&\/=]+$`,
	)

	providers = map[int]string{
		youTube:    "Youtube",
		soundCloud: "SoundCloud",
		vimeo:      "Vimeo",
	}
	embedPatterns = [...]struct {
		typ  int
		patt *regexp.Regexp
	}{
		{
			youTube,
			regexp.MustCompile(`https?:\/\/(?:[^\.]+\.)?youtube\.com\/watch\/?\?(?:.+&)?v=([^&]+)`),
		},
		{
			youTube,
			regexp.MustCompile(`https?:\/\/(?:[^\.]+\.)?(?:youtu\.be|youtube\.com\/embed)\/([a-zA-Z0-9_-]+)`),
		},
		{
			soundCloud,
			regexp.MustCompile(`https?:\/\/soundcloud.com\/.*`),
		},
		{
			vimeo,
			regexp.MustCompile(`https?:\/\/(?:www\.)?vimeo\.com\/.+`),
		},
	}
)

type bodyContext struct {
	state struct { // Body parser state
		spoiler, quote bool
		iDice          int
	}
	common.Post
	OP uint64
	quicktemplate.Writer
}

// Render the text body of a post
func streambody(w *quicktemplate.Writer, p common.Post, op uint64) {
	c := bodyContext{
		Post:   p,
		OP:     op,
		Writer: *w,
	}

	lines := strings.Split(c.Body, "\n")
	if c.Editing {
		for i := 0; i < len(lines)-1; i++ {
			c.parseTerminatedLine(lines[i])
		}
		c.parseOpenLine(lines[len(lines)-1])
	} else {
		for _, line := range lines {
			c.parseTerminatedLine(line)
		}
	}
}

// Parse a line that is no longer being edited
func (c *bodyContext) parseTerminatedLine(line string) {
	// For hiding redundant newlines using CSS
	if len(line) == 0 {
		c.N().S("<br>")
		return
	}

	c.initLine(line[0])

	if line[0] == '#' {
		if m := commandRegexp.FindStringSubmatch(line); m != nil {
			c.parseCommands(string(m[1]))
			c.terminateTags(true)
			return
		}
	}

	c.parseSpoilers(line, (*c).parseFragment)
	c.terminateTags(true)
}

// Open a new line container and check for quotes
func (c *bodyContext) initLine(first byte) {
	c.state.spoiler = false
	c.state.quote = false

	c.N().S("<span>")
	if first == '>' {
		c.N().S("<em>")
		c.state.quote = true
	}
}

// Injects spoiler tags and calls fn on the remaining parts
func (c *bodyContext) parseSpoilers(frag string, fn func(string)) {
	for {
		i := strings.Index(frag, "**")
		if i != -1 {
			fn(frag[:i])
			if c.state.spoiler {
				c.N().S("</del>")
			} else {
				c.N().S("<del>")
			}
			c.state.spoiler = !c.state.spoiler
			frag = frag[i+2:]
		} else {
			fn(frag)
			break
		}
	}
}

// Parse a line fragment
func (c *bodyContext) parseFragment(frag string) {
	for i, word := range strings.Split(frag, " ") {
		if i != 0 {
			c.N().S(` `)
		}
		if len(word) == 0 {
			continue
		}
		if word[0] == '>' {
			if m := linkRegexp.FindStringSubmatch(word); m != nil {
				// Post links
				c.parsePostLink(m)
				continue
			} else if m := referenceRegexp.FindStringSubmatch(word); m != nil {
				// Internal and custom reference URLs
				c.parseReference(m)
				continue
			}
		} else {
			// Generic HTTP(S) URLs and magnet links
			match := false
			// Checking the first byte is much cheaper than a function call. Do
			// that first, as most cases won't match.
			switch word[0] {
			case 'h':
				match = strings.HasPrefix(word, "http")
			case 'm':
				match = strings.HasPrefix(word, "magnet:?")
			}
			if match {
				c.parseURL(word)
				continue
			}
		}
		c.E().S(word)
	}
}

// Parse a potential link to a post
func (c *bodyContext) parsePostLink(m []string) {
	if c.Links == nil {
		c.N().S(m[0])
		return
	}

	id, _ := strconv.ParseUint(string(m[2]), 10, 64)
	verified, ok := c.Links[id]
	if !ok {
		c.N().S(m[0])
		return
	}

	if len(m[1]) != 0 { // Write extra quotes
		c.N().S(m[1])
	}
	streampostLink(
		&c.Writer,
		id,
		verified.OP,
		verified.Board,
		verified.OP != c.OP,
	)
}

// Parse internal or customly set reference URL
func (c *bodyContext) parseReference(m []string) {
	var (
		m2   = string(m[2])
		href string
	)
	if config.IsBoard(m2) {
		href = fmt.Sprintf("/%s/", m2)
	} else if href = config.Get().Links[m2]; href != "" {
	} else {
		c.N().S(m[0])
		return
	}

	if len(m[1]) != 0 {
		c.N().S(m[1])
	}
	c.newTabLink(href, fmt.Sprintf(">>>/%s/", string(m[2])))
}

// Format and anchor link that opens in a new tab
func (c *bodyContext) newTabLink(href, text string) {
	c.N().S(`<a href="`)
	c.E().S(href)
	c.N().S(`" target="_blank">`)
	c.E().S(text)
	c.N().S(`</a>`)
}

// Parse generic URLs and magnet links
func (c *bodyContext) parseURL(bit string) {
	s := string(bit)
	switch {
	case !urlRegexp.MatchString(bit):
		c.E().S(bit)
	case c.parseEmbeds(bit):
	case bit[0] == 'm': // Don't open a new tab for magnet links
		s = html.EscapeString(s)
		c.N().S(`<a href="`)
		c.N().S(s)
		c.N().S(`">`)
		c.N().S(s)
		c.N().S(`</a>`)
	default:
		c.newTabLink(s, s)
	}
}

// Parse select embeddable URLs. Returns, if any found.
func (c *bodyContext) parseEmbeds(s string) bool {
	for _, t := range embedPatterns {
		if !t.patt.MatchString(s) {
			continue
		}

		c.N().S(`<em><a class="embed" target="_blank" data-type="`)
		c.N().D(t.typ)
		c.N().S(`" href="`)
		c.E().S(s)
		c.N().S(`">[`)
		c.N().S(providers[t.typ])
		c.N().S(`] ???</a></em>`)

		return true
	}
	return false
}

// Parse a hash command
func (c *bodyContext) parseCommands(bit string) {
	// Guard against invalid dice rolls
	invalid := c.Commands == nil ||
		c.state.iDice > len(c.Commands)-1 ||
		c.Commands[c.state.iDice].Val == nil
	if invalid {
		c.writeInvalidCommand(bit)
		return
	}

	// TODO: Sycnwatch

	inner := new(bytes.Buffer)
	switch bit {
	case "flip", "8ball", "pyu", "pcount":
		fmt.Fprint(inner, c.Commands[c.state.iDice].Val)
		c.state.iDice++
	default:
		// Validate dice
		m := diceRegexp.FindStringSubmatch(bit)
		if m[1] != "" {
			if rolls, err := strconv.Atoi(m[1]); err != nil || rolls > 10 {
				c.writeInvalidCommand(bit)
				return
			}
		}
		if sides, err := strconv.Atoi(m[2]); err != nil || sides > 100 {
			c.writeInvalidCommand(bit)
			return
		}

		// Cast []interface to []uint16
		uncast := c.Commands[c.state.iDice].Val.([]interface{})
		rolls := make([]uint16, len(uncast))
		for i := range rolls {
			rolls[i] = uint16(uncast[i].(float64))
		}

		c.state.iDice++
		var sum uint
		for i, roll := range rolls {
			if i != 0 {
				inner.WriteString(" + ")
			}
			sum += uint(roll)
			inner.WriteString(strconv.FormatUint(uint64(roll), 10))
		}
		if len(rolls) > 1 {
			fmt.Fprintf(inner, " = %d", sum)
		}
	}

	c.N().S(`<strong>#`)
	c.N().S(bit)
	c.N().S(` (`)
	c.N().S(inner.String())
	c.N().S(`)</strong>`)
}

// If command validation failed, simply write the string
func (c *bodyContext) writeInvalidCommand(s string) {
	c.N().S("#")
	c.N().S(s)
}

// Close any open HTML tags
func (c *bodyContext) terminateTags(newLine bool) {
	if c.state.spoiler {
		c.N().S("</del>")
	}
	if c.state.quote {
		c.N().S("</em>")
	}
	if newLine {
		c.N().S("<br>")
	}
	c.N().S("</span>")
}

// Parse a line that is still being edited
func (c *bodyContext) parseOpenLine(line string) {
	if len(line) == 0 {
		c.N().S("<span></span>")
		return
	}
	c.initLine(line[0])
	c.parseSpoilers(line, func(s string) {
		c.E().S(s)
	})
	c.terminateTags(false)
}

File: ts/posts/render/body.ts
/**
 * Post formatting renderer. Uses marked with custom rules.
 */

import * as marked from "marked"
import { renderPostLink } from './etc'
import { PostData, PostLink, TextState } from '../../common'
import { escape, makeAttrs } from '../../util'
import { parseEmbeds } from "../embed"

type AnyClass = { new(...args: any[]): any }
type Renderer = (post: PostData) => string

function noop() {}
;(noop as any).exec = noop

// Create renderer instance.
function makeRenderer(): Renderer {
  // Custom Lexer.
  class Lexer extends ((marked as any).Lexer as AnyClass) {
    constructor(options?: any) {
      super(options)
      // TODO(Kagami): Remove def from regexes?
      Object.assign(this.rules, {
        code: noop,
        hr: noop,
        heading: noop,
        lheading: noop,
        blockquote: /^( *>[^>\n][^\n]*)/,
        def: noop,
      })
    }
  }

  // Custom InlineLexer.
  class InlineLexer extends ((marked as any).InlineLexer as AnyClass) {
    constructor(links: any, options: any, post: PostData) {
      // XXX(Kagami): Inject post link logic via hardcoded link defs.
      // Hacky, but unfortunately marked can't be easily customized.
      links[""] = {href: "post-link", title: ""}
      super(links, options)
      this.post = post
      Object.assign(this.rules, {
        link: noop,
        reflink: /^>>\d+()/,
        nolink: noop,
      })
    }
    outputLink(cap: any, link: any) {
      if (link.href === "post-link") {
        return parsePostLink(cap, this.post.links, this.post.op)
      }
      return super.outputLink(cap, link)
    }
  }

  // Custom Parser.
  class Parser extends (marked as any).Parser {
    parse(src: any, post: PostData) {
      this.inline = new InlineLexer(src.links, this.options, post)
      this.tokens = src.reverse()

      let out = ""
      while (this.next()) {
        out += this.tok()
      }

      return out
    }
  }

  // Custom Renderer.
  class Renderer extends marked.Renderer {
    blockquote(quote: string): string {
      return "<blockquote>&gt; " + quote + "</blockquote>"
    }
    // paragraph(text: string): string {
    //   return text
    // }
  }

  // Set defaults.
  marked.setOptions({
    // gfm: true,
    tables: false,
    breaks: true,
    // pedantic: false,
    sanitize: true,  // Very important!
    // sanitizer: null,
    // mangle: true,
    // smartLists: false,
    // silent: false,
    // highlight: null,
    // langPrefix: 'lang-',
    // smartypants: false,
    // headerPrefix: '',
    renderer: new Renderer(),
    // xhtml: false
  })

  // Resulting render function.
  return function(post) {
    const lexer = new Lexer()
    const tokens = lexer.lex(post.body)
    const parser = new Parser()
    return parser.parse(tokens, post)
  }
}

const renderer = makeRenderer()

// Render Markdown-like post body to sanitized HTML.
export function render(post: PostData): string {
  return renderer(post)
}

// Verify and render a link to other posts.
function parsePostLink(m: string[], links: PostLink[], thread: number): string {
  if (!links) return m[0]
  const id = +m[0].slice(2)
  const link = links.find(l => l[0] === id)
  if (!link) return m[0]
  return renderPostLink(id, link[1], thread)
}

// URLs supported for linkification
const urlPrefixes = {
  'h': "http",
  'm': "magnet:?",
  'i': "irc",
  'f': "ftp",
  'b': "bitcoin",
}

// Render the text body of a post
export default function(data: PostData): string {
  const state: TextState = data.state = {
    spoiler: false,
    quote: false,
    lastLineEmpty: false,
    code: false,
    haveSyncwatch: false,
    iDice: 0,
  }
  let html = ""

  const fn = parseTerminatedLine,
    lines = data.body.split("\n"),
    last = lines.length - 1
  for (let i = 0; i < lines.length; i++) {
    const l = lines[i]

    // Prevent successive empty lines
    if (!l.length) {
      // Don't break, if body ends with newline
      if (!state.lastLineEmpty && i !== last) {
        html += "<br>"
      }
      state.lastLineEmpty = true
      state.quote = false
      continue
    }

    html += initLine(l, state)
      + fn(l, data)
      + terminateTags(state, i != last)
  }

  return html
}

// Parse a single line, that is no longer being edited
function parseTerminatedLine(line: string, data: PostData): string {
  return parseCode(line, data.state, frag =>
    parseFragment(frag, data))
}

// Detect code tags
function parseCode(
  frag: string,
  state: TextState,
  fn: (frag: string) => string,
): string {
  let html = ""
  while (true) {
    const i = frag.indexOf("``")
    if (i !== -1) {
      html += formatCode(frag.slice(0, i), state, fn)
      frag = frag.substring(i + 2)
      state.code = !state.code
    } else {
      html += formatCode(frag, state, fn)
      break
    }
  }
  return html
}

function formatCode(
  frag: string,
  state: TextState,
  fn: (frag: string) => string,
): string {
  return parseSpoilers(frag, state, fn)
}

// Injects spoiler tags and calls fn on the remaining parts
function parseSpoilers(
  frag: string,
  state: TextState,
  fn: (frag: string) => string,
): string {
  let html = ""
  while (true) {
    const i = frag.indexOf("**")
    if (i !== -1) {
      html += fn(frag.slice(0, i))
      if (state.quote) {
        html += "</em>"
      }
      html += `<${state.spoiler ? '/' : ''}del>`
      if (state.quote) {
        html += "<em>"
      }

      state.spoiler = !state.spoiler
      frag = frag.substring(i + 2)
    } else {
      html += fn(frag)
      break
    }
  }
  return html
}

// Open a new line container and check for quotes
function initLine(line: string, state: TextState): string {
  let html = ""
  state.quote = state.lastLineEmpty = false
  if (line[0] === ">") {
    state.quote = true
    html += "<em>"
  }
  if (state.spoiler) {
    html += "<del>"
  }
  return html
}

// Close all open tags at line end
function terminateTags(state: TextState, newLine: boolean): string {
  let html = ""
  if (state.spoiler) {
    html += "</del>"
  }
  if (state.quote) {
    html += "</em>"
  }
  if (newLine) {
    html += "<br>"
  }
  return html
}

// Parse a line fragment
function parseFragment(frag: string, data: PostData): string {
  let html = ""
  const words = frag.split(" ")
  for (let i = 0; i < words.length; i++) {
    if (i !== 0) {
      html += " "
    }

    // Split leading and trailing punctuation, if any
    const [leadPunct, word, trailPunct] = splitPunctuation(words[i])
    if (leadPunct) {
      html += leadPunct
    }
    if (!word) {
      if (trailPunct) {
        html += trailPunct
      }
      continue
    }

    let m: RegExpMatchArray,
      matched = false
    switch (word[0]) {
      case ">":
        // Post links
        m = word.match(/^>>(>*)(\d+)$/)
        if (m) {
          html += parsePostLink(m, data.links, data.op)
          matched = true
        }
        break
      case "#": // Hash commands
        m = word.match(/^#(flip|\d*d\d+|pyu|pcount|sw(?:\d+:)?\d+:\d+(?:[+-]\d+)?)$/)
        if (m) {
          html += parseCommand(m[1], data)
          matched = true
        }
        break
      default:
        // Generic HTTP(S) URLs, magnet links and embeds
        // Checking the first byte is much cheaper than a function call.
        // Do that first, as most cases won't match.
        const pre = urlPrefixes[word[0]]
        if (pre && word.startsWith(pre)) {
          html += parseURL(word)
          matched = true
        }
    }

    if (!matched) {
      html += escape(word)
    }
    if (trailPunct) {
      html += trailPunct
    }
  }

  return html
}

// Render and anchor link that opens in a new tab
function newTabLink(href: string, text: string): string {
  const attrs = {
    rel: "noreferrer",
    href: escape(href),
    target: "_blank",
  }
  return `<a ${makeAttrs(attrs)}>${escape(text)}</a>`
}

// Parse generic URLs and embed, if applicable
function parseURL(bit: string): string {
  const embed = parseEmbeds(bit)
  if (embed) {
    return embed
  }

  try {
    new URL(bit) // Will throw, if invalid URL
    if (bit[0] == "m") { // Don't open a new tab for magnet links
      bit = escape(bit)
      return bit.link(bit)
    }
    return newTabLink(bit, bit)
  } catch (e) {
    return escape(bit)
  }
}

// Parse a hash command
function parseCommand(bit: string, { commands, state }: PostData): string {
  // Guard against invalid dice rolls and parsing lines in the post form
  if (!commands || !commands[state.iDice]) {
    return "#" + bit
  }

  let inner: string
  switch (bit) {
    case "flip":
    case "pyu":
    case "pcount":
      inner = commands[state.iDice++].val.toString()
      break
    default:
      if (bit.startsWith("sw")) {
        return formatSyncwatch(bit, commands[state.iDice++].val, state)
      }

      // Validate dice
      const m = bit.match(/^(\d*)d(\d+)$/)
      if (parseInt(m[1]) > 10 || parseInt(m[2]) > 100) {
        return "#" + bit
      }

      const rolls = commands[state.iDice++].val as number[]
      inner = ""
      let sum = 0
      for (let i = 0; i < rolls.length; i++) {
        if (i) {
          inner += " + "
        }
        sum += rolls[i]
        inner += rolls[i]
      }
      if (rolls.length > 1) {
        inner += " = " + sum
      }
  }

  return `<strong>#${bit} (${inner})</strong>`
}

// Format a synchronized time counter
function formatSyncwatch(bit: string, val: number[], state: TextState): string {
  state.haveSyncwatch = true
  const attrs = {
    class: "embed syncwatch",
    "data-hour": val[0].toString(),
    "data-min": val[1].toString(),
    "data-sec": val[2].toString(),
    "data-start": val[3].toString(),
    "data-end": val[4].toString()
  }
  return `<em><strong ${makeAttrs(attrs)}>syncwatch</strong></em>`
}

// Splits off one byte of leading and trailing punctuation, if any, and returns
// the 3 split parts. If there is no edge punctuation, the respective string
// is empty.
function splitPunctuation(word: string): [string, string, string] {
  const re: [string, string, string] = ["", word, ""]
  re[1] = word

  // Split leading
  if (re[1].length < 2) {
    return re
  }
  if (isPunctuation(re[1][0])) {
    re[0] = re[1][0]
    re[1] = re[1].slice(1)
  }

  // Split trailing
  const l = re[1].length
  if (l < 2) {
    return re
  }
  if (isPunctuation(re[1][l - 1])) {
    re[2] = re[1][l - 1]
    re[1] = re[1].slice(0, -1)
  }

  return re
}

// Return if b is a punctuation byte
function isPunctuation(b: string): boolean {
  switch (b) {
    case '!':
    case '"':
    case '\'':
    case '(':
    case ')':
    case ',':
    case '-':
    case '.':
    case ':':
    case ';':
    case '?':
    case '[':
    case ']':
      return true
    default:
      return false
  }
}
